File: lab5/haskel/totient.hs
import Data.List (nub)

factorize :: Int -> Int -> [Int]
factorize 1 _ = []
factorize n divisor
    | n `mod` divisor == 0 = divisor : factorize (n `div` divisor) divisor
    | otherwise            = factorize n (divisor + 1)

prime_factors :: Int -> [Int]
prime_factors n
    | n <= 1    = []
    | otherwise = factorize n 2

coprime :: Int -> [Int]
coprime n = [k | k <- [1..n], gcd n k == 1]

totient1 :: Int -> Int
totient1 n = length $ coprime n

totient2 :: Int -> Int
totient2 n =
    let factors = prime_factors n
        uniqueFactors = nub factors  -- removeDuplicates
    in 
        foldl (\acc p -> acc `div` p * (p - 1)) n uniqueFactors

File: lab5/prolog/totient.pl
gcd(A, 0, A).
gcd(A, B, GCD) :-
    B > 0,
    C is A mod B,
    gcd(B, C, GCD).

coprime(X, Y) :-
    gcd(X, Y, GCD),
    GCD =:= 1.

totientUtil(_, 1, T, T) :- !.
totientUtil(N, M, Acc, T) :-
    M > 1,
    coprime(N, M),
    NewAcc is Acc + 1,
    NewM is M - 1,
    totientUtil(N, NewM, NewAcc, T).
totientUtil(N, M, Acc, T) :-
    M > 1,
    \+ coprime(N, M),
    NewM is M - 1,
    totientUtil(N, NewM, Acc, T).

totient(1, 1).
totient(N, T) :-
    N > 1,
    totientUtil(N, N, 1, T).